### PR TITLE
Pass reset

### PR DESCRIPTION
**Pull request by Tim-K-DFW**

### CODE REVIEW

**Comment by matthew-ta** [comment]
Looks like you have some merge conflicts here to fix.

File: app/models/user.rb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 class User < ActiveRecord::Base
   validates :username, presence: true, uniqueness: true, length: {minimum: 4}
-  validates :password, presence: true, length: {minimum: 3}
+  # validates :password, presence: true, length: {minimum: 3}
```
Comment (matthew-ta) [inline app/models/user.rb]: Remove if not using

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -22,9 +22,47 @@ def create
...
 
+  def request_reset
+  end
+
+  def send_reset_link
+    user = User.where(email: params[:email]).first
+    if user
+      scheme = request.env['rack.url_scheme']
+      host = request.env['HTTP_HOST']
+      link = user.generate_reset_link(scheme, host)
```
Comment (matthew-ta) [inline app/controllers/users_controller.rb]: These two lines IMO should be be done in the mailer or by using a `url` helper.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -22,9 +22,47 @@ def create
...
+      AppMailer.send_password_reset_link(user, link).deliver
+      render 'confirm_password_reset'
+    else
+      flash[:danger] = 'If you forgot your email address as well, you can register again.'
+      render 'request_reset'
+    end
+  end
+
+  def enter_new_password
```
Comment (matthew-ta) [inline app/controllers/users_controller.rb:40]: In my opinion I feel like this is moving too far away from the standard CRUD model. I would suggest moving to a `passwords_controller` where you can use the standard CRUD method. Maybe `passwords_resets_controller` might be a better name.

File: app/models/user.rb [hunk 4]
```diff
@@ -24,4 +24,10 @@ def follows?(another_user)
   def can_follow?(another_user)
     !follows?(another_user) && self != another_user
   end
+
+  def generate_reset_link(scheme=nil, host=nil)
+    token = SecureRandom.urlsafe_base64
+    self.update(token: token)
+    "#{scheme}://#{host}/reset_password/#{self.token}"
```
Comment (matthew-ta) [inline app/models/user.rb]: I don't think you should return the full URL here, just saving the token should be fine.

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (Tim-K-DFW) [inline app/controllers/users_controller.rb]: Any chance you could show how or which one to use? I tried to use it inside a model but couldn't make it access environment. I feel rails must have something simple and elegant for this, but what?

**Comment by Tim-K-DFW** [comment]
What's the best way to merge it without messing up this branch? I had one commit on Master which I don't need. When tried to revert from console and then push to master, it created 2nd commit instead... Just need to 'clean up' Master to the point where I created the branch.

Agree it's pretty hairy, will be more careful with commits from now on.

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (matthew-ta) [inline app/controllers/users_controller.rb]: It should be something like `reset_link_url("token")`. This from an app I am working on right now - http://cl.ly/image/3m361g2p3Y25

**Pull request merged**